Persist on init if state changed

andrepimenta opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

andrepimenta committed:
Persist on init if state changed

Gudahtt commented on app/core/EngineService/EngineService.ts:
```diff
@@ -197,26 +209,59 @@ export class EngineService {
 
+          // If initialState provided, check if state changed during Engine.init()
+          // Controllers preserve referential equality, so shallow comparison is sufficient
+          if (initialState) {
```
We will always want to persist if there is no initial state.
 
 Maybe easiest to remove this condition, and update the line below to:
 
 ```
             const initialControllerState = initialState?.[controllerName];
 ```
 
 so that it will resolve as `undefined` for new controllers/new installs and trigger persistence

andrepimenta committed:
Update app/core/EngineService/EngineService.ts

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

andrepimenta committed:
fix(EngineService): persist controller state changes during Engine.init()

Controllers can modify state during Engine.init() (e.g., SnapController
granting permissions to PermissionController). These changes were not
being persisted because persistence subscriptions couldn't capture events
emitted during controller construction.

This fix compares controller state before and after Engine.init() using
1-level deep referential equality. Controllers preserve referential
equality for unchanged properties (via Immer), so if a property reference
differs, the state changed and needs persistence.

Changes:
- Compare filteredState properties against initialState at 1-level deep
- Handle new installs/controllers where initialState is undefined
- Only persist if there's actual state to save (non-empty filteredState)
- Reuse the same debounced persistController function for consistency
